server/src/main/java/org/opensearch/index/IndexService.java

```diff
@@ -516,6 +517,9 @@ public synchronized IndexShard createShard(
             }
 
             Directory directory = directoryFactory.newDirectory(this.indexSettings, path);
+            if (directory instanceof CompositeDirectory) {
+                ((CompositeDirectory) directory).setRemoteDirectory(remoteDirectory);
```

Comment:
Instead of using the setter model, can we construct the composite directory by providing both local and remote directory as input to it ? Composite Directory will be used for all remote backed indices. From what I see, is that CompositeDirectory probably will not follow the DirectoryFactory model which other directory types are following here, instead would be an internal class to compose 2 directories created by separate factories?

Reply:
Agree @sohami 
Providing both local and remote directory as input to the Composite Directory makes more sense.
Also having a DirectoryFactory doesn't make much sense either.
Have done so in the latest revision.